File: backend/go-chi-hive/internal/cells/article/service.go
package article

import (
	"context"
	"time"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hive/internal/shared/errors"
)

type service struct {
	repo         Repository
	userProvider UserProvider
}

func NewService(repo Repository, userProvider UserProvider) Service {
	return &service{
		repo:         repo,
		userProvider: userProvider,
	}
}

func (s *service) CreateArticle(ctx context.Context, cmd CreateArticleCommand) (*Article, error) {
	author, err := s.userProvider.GetUser(ctx, cmd.AuthorID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if author == nil {
		return nil, errors.NewResourceNotFound("User", "id", cmd.AuthorID)
	}

	// Check if article with same title already exists
	existingArticle, err := s.repo.GetByTitle(ctx, cmd.Title, nil)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if existingArticle != nil {
		return nil, errors.NewAlreadyExistsError("Article with this title already exists")
	}

	slug := Slugify(cmd.Title)

	// Check if article with same slug already exists
	existingArticle, err = s.repo.GetBySlug(ctx, slug, nil)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if existingArticle != nil {
		return nil, errors.NewAlreadyExistsError("Article with this slug already exists")
	}

	article := &Article{
		Slug:           slug,
		Title:          cmd.Title,
		Description:    cmd.Description,
		Body:           cmd.Body,
		TagList:        cmd.TagList,
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
		Favorited:      false,
		FavoritesCount: 0,
		Author: Author{
			Username:  author.Username,
			Bio:       author.Bio,
			Image:     author.Image,
			Following: false,
		},
	}

	if err := s.repo.Create(ctx, article, author.ID); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return article, nil
}

func (s *service) GetArticle(ctx context.Context, query GetArticleQuery) (*Article, error) {
	article, err := s.repo.GetBySlug(ctx, query.Slug, query.ObserverID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, errors.NewResourceNotFound("Article", "slug", query.Slug)
	}

	return article, nil
}

func (s *service) GetArticles(ctx context.Context, query GetArticlesQuery) ([]*Article, int, error) {
	articles, count, err := s.repo.GetArticles(ctx, query)
	if err != nil {
		return nil, 0, errors.NewInternalError(err.Error())
	}
	return articles, count, nil
}

func (s *service) GetFeed(ctx context.Context, query GetFeedQuery) ([]*Article, int, error) {
	articles, count, err := s.repo.GetFeed(ctx, query.UserID, query.Limit, query.Offset)
	if err != nil {
		return nil, 0, errors.NewInternalError(err.Error())
	}
	return articles, count, nil
}

func (s *service) UpdateArticle(ctx context.Context, cmd UpdateArticleCommand) (*Article, error) {
	// Get article and check if it exists
	article, err := s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, errors.NewResourceNotFound("Article", "slug", cmd.Slug)
	}

	// Check if user is author
	author, err := s.userProvider.GetUserByUsername(ctx, article.Author.Username)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if author == nil || author.ID != cmd.UserID {
		return nil, errors.NewForbiddenError("you are not the author of this article")
	}

	// Define duplicate check function
	checkDuplicate := func(title, slug string) error {
		// Check title
		existing, err := s.repo.GetByTitle(ctx, title, nil)
		if err != nil {
			return errors.NewInternalError(err.Error())
		}
		if existing != nil && existing.ID != article.ID {
			return errors.NewAlreadyExistsError("Article with this title already exists")
		}

		// Check slug
		existing, err = s.repo.GetBySlug(ctx, slug, nil)
		if err != nil {
			return errors.NewInternalError(err.Error())
		}
		if existing != nil && existing.ID != article.ID {
			return errors.NewAlreadyExistsError("Article with this slug already exists")
		}
		return nil
	}

	if err := article.Update(cmd.Title, cmd.Description, cmd.Body, checkDuplicate); err != nil {
		return nil, err
	}

	if err := s.repo.Update(ctx, article); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return article, nil
}

func (s *service) DeleteArticle(ctx context.Context, cmd DeleteArticleCommand) error {
	article, err := s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
	if err != nil {
		return errors.NewInternalError(err.Error())
	}
	if article == nil {
		return errors.NewResourceNotFound("Article", "slug", cmd.Slug)
	}

	author, err := s.userProvider.GetUserByUsername(ctx, article.Author.Username)
	if err != nil {
		return errors.NewInternalError(err.Error())
	}
	if author == nil || author.ID != cmd.UserID {
		return errors.NewForbiddenError("you are not the author of this article")
	}

	if err := s.repo.Delete(ctx, article.ID); err != nil {
		return errors.NewInternalError(err.Error())
	}

	return nil
}

func (s *service) FavoriteArticle(ctx context.Context, cmd FavoriteArticleCommand) (*Article, error) {
	article, err := s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, errors.NewResourceNotFound("Article", "slug", cmd.Slug)
	}

	if err := s.repo.Favorite(ctx, article.ID, cmd.UserID); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
}

func (s *service) UnfavoriteArticle(ctx context.Context, cmd UnfavoriteArticleCommand) (*Article, error) {
	article, err := s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
	if err != nil {
		return nil, errors.NewInternalError(err.Error())
	}
	if article == nil {
		return nil, errors.NewResourceNotFound("Article", "slug", cmd.Slug)
	}

	if err := s.repo.Unfavorite(ctx, article.ID, cmd.UserID); err != nil {
		return nil, errors.NewInternalError(err.Error())
	}

	return s.repo.GetBySlug(ctx, cmd.Slug, &cmd.UserID)
}

func (s *service) GetTags(ctx context.Context) ([]Tag, error) {
	return s.repo.FindAllTags(ctx)
}

File: backend/python-fastapi-vertical/app/features/article/service.py
from datetime import UTC
from typing import Optional

from features.article.domain import Article, Author, slugify
from features.article.repository import (
    ArticlesList,
    create,
    find_by_slug,
    find_by_slug_for_author,
    find_all,
    find_feed,
    update,
    delete,
    favorite,
    unfavorite,
    find_all_tags,
)
from features.user.repository import find_by_id as find_user_by_id, find_by_username
from shared.errors.app_error import (
    new_already_exists_error,
    new_forbidden_error,
    new_resource_not_found,
)


async def create_article(
    author_id: int,
    title: str,
    description: str,
    body: str,
    tag_list: Optional[list[str]] = None,
) -> Article:
    slug = slugify(title)
    existing = await find_by_slug(slug)
    if existing:
        raise new_already_exists_error("Article with this title already exists")

    user = await find_user_by_id(author_id)
    if not user:
        raise new_resource_not_found("User", "id", author_id)

    author = Author(
        username=user.username,
        bio=user.bio,
        image=user.image,
        following=False,
    )

    from datetime import datetime

    article = Article(
        id=0,
        slug=slug,
        title=title,
        description=description,
        body=body,
        tag_list=tag_list or [],
        created_at=datetime.now(UTC),
        updated_at=datetime.now(UTC),
        favorited=False,
        favorites_count=0,
        author=author,
    )

    await create(article, author_id)
    return article


async def get_article(slug: str, observer_id: Optional[int] = None) -> Article:
    article = await find_by_slug(slug, observer_id)
    if not article:
        raise new_resource_not_found("Article", "slug", slug)
    return article


async def update_article(
    slug: str,
    user_id: int,
    title: Optional[str] = None,
    description: Optional[str] = None,
    body: Optional[str] = None,
) -> Article:
    result = await find_by_slug_for_author(slug)
    if not result:
        raise new_resource_not_found("Article", "slug", slug)

    article, author_id = result

    if author_id != user_id:
        raise new_forbidden_error("You are not allowed to update this article")

    if title is not None:
        new_slug = slugify(title)
        if new_slug != article.slug:
            existing = await find_by_slug(new_slug)
            if existing:
                raise new_already_exists_error("Article with this title already exists")

    article.update(title=title, description=description, body=body)
    await update(article)
    return article


async def delete_article(slug: str, user_id: int) -> None:
    result = await find_by_slug_for_author(slug)
    if not result:
        raise new_resource_not_found("Article", "slug", slug)

    article, author_id = result

    if author_id != user_id:
        raise new_forbidden_error("You are not allowed to delete this article")

    await delete(article.id)


async def favorite_article(slug: str, user_id: int) -> Article:
    article = await find_by_slug(slug, user_id)
    if not article:
        raise new_resource_not_found("Article", "slug", slug)

    await favorite(article.id, user_id)

    refreshed = await find_by_slug(slug, user_id)
    return refreshed


async def unfavorite_article(slug: str, user_id: int) -> Article:
    article = await find_by_slug(slug, user_id)
    if not article:
        raise new_resource_not_found("Article", "slug", slug)

    await unfavorite(article.id, user_id)

    refreshed = await find_by_slug(slug, user_id)
    return refreshed


async def get_tags() -> list[str]:
    return await find_all_tags()


async def get_articles(
    tag: Optional[str] = None,
    author: Optional[str] = None,
    favorited: Optional[str] = None,
    limit: int = 20,
    offset: int = 0,
    observer_id: Optional[int] = None,
) -> ArticlesList:
    return await find_all(
        tag=tag,
        author=author,
        favorited=favorited,
        limit=limit,
        offset=offset,
        observer_id=observer_id,
    )


async def get_feed(
    user_id: int,
    limit: int = 20,
    offset: int = 0,
) -> ArticlesList:
    return await find_feed(
        user_id=user_id,
        limit=limit,
        offset=offset,
    )